src/tokenizer/generic_sql.cpp

```diff
@@ -0,0 +1,201 @@
+        advance(number_str.size() - 1);
+    } else {
+        // If it's not a number, it must be an operator
+        token.str = substr(token.index, 1);
+        token.type = sql_token_type::binary_operator;
+    }
+
+    emplace_token(token);
+}
+
+std::vector<sql_token> generic_sql_tokenizer::tokenize_impl()
+{
+    for (; !eof(); advance()) {
+        auto c = peek();
+        // TODO use an array of characters or a giant switch?
+        if (ddwaf::isalpha(c) || c == '_' ||
+            static_cast<unsigned char>(c) > 0x7f) { // Command or identifier
+            tokenize_command_operator_or_identifier();
+        } else if (ddwaf::isdigit(c)) {
+            tokenize_number();
+        } else if (c == '"') {
+            // For compatibility, double-quoted strings are always considered
+            // identifiers.
+            tokenize_string('"', sql_token_type::identifier);
+        } else if (c == '\'') { // Single-quoted string
+            tokenize_string('\'', sql_token_type::single_quoted_string);
+        } else if (c == '`') { // Backtick-quoted string
+            tokenize_string('`', sql_token_type::back_quoted_string);
+        } else if (c == '(') {
+            add_token(sql_token_type::parenthesis_open);
+        } else if (c == ')') {
+            add_token(sql_token_type::parenthesis_close);
+        } else if (c == '.') {
+            add_token(sql_token_type::dot);
+        } else if (c == ',') {
+            add_token(sql_token_type::comma);
+        } else if (c == '?') {
+            add_token(sql_token_type::questionmark);
+        } else if (c == '*') {
+            add_token(sql_token_type::asterisk);
+        } else if (c == ';') {
+            add_token(sql_token_type::query_end);
+        } else if (c == '/') {
+            tokenize_inline_comment_or_operator();
+        } else if (c == '-') {
+            tokenize_eol_comment_operator_or_number();
+        } else if (c == '#') {
+            tokenize_eol_comment();
+        } else if (c == '+') {
+            tokenize_operator_or_number();
+        } else if (c == '!' || c == '>') {
+            add_token(sql_token_type::binary_operator, next() == '=' ? 2 : 1);
+        } else if (c == '<') {
+            auto n = next();
+            if (n == '=' || n == '>') {
+                add_token(sql_token_type::binary_operator, 2);
+            } else {
+                add_token(sql_token_type::binary_operator);
+            }
+        } else if (c == '=' || c == '%') {
+            add_token(sql_token_type::binary_operator);
+        } else if (c == '|') {
+            if (next() == '|') {
+                add_token(sql_token_type::binary_operator, 2);
+            } else {
+                add_token(sql_token_type::bitwise_operator);
+            }
+        } else if (c == '&' || c == '^' || c == '~') {
+            add_token(sql_token_type::bitwise_operator);
+        } else if (c == ':') {
+            add_token(sql_token_type::colon);
+        }
```

Comment:
`else`?... I reckon you're expecting whitespace here, but if it's not you should error.

Reply:
The tokenizers are non-exhaustive by design, if we find characters we don't understand we ignore them, hence the lack of `else`.

Reply:
> The tokenizers are non-exhaustive by design, if we find characters we don't understand we ignore them, hence the lack of `else`.

I think that's a bit of a dangerous option. There could be some unrecognized structure changing the context (comment, type string, etc.).

Reply:
After some thought, I've changed the tokenizers to fail and log if we find an unexpected character.